llm/llm.go

```diff
@@ -50,92 +50,88 @@ func New(model string, adapters, projectors []string, opts api.Options) (LLM, er
 	}
 
 	vram, _ := gpu.CheckVRAM()
-	size := ggml.Size
+	info := gpu.GetGPUInfo()
+
+	mem := int64(396 * 1024 * 1024)
+	for _, projector := range projectors {
+		mem += projectorMemoryRequirements(projector)
+	}
 
-	// fp16 k,v matrices require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each * 2 key and value
-	kv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.KV().BlockCount()) * int64(ggml.KV().EmbeddingLength()) * int64(ggml.KV().HeadCountKV()) / int64(ggml.KV().HeadCount())
+	// fp16 k,v = (1 (k) + 1 (v)) * sizeof(float16) * n_ctx * n_layer * n_embd / n_head * n_head_kv
+	kv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.KV().BlockCount()) * int64(ggml.KV().EmbeddingLength()) / int64(ggml.KV().HeadCount()) * int64(ggml.KV().HeadCountKV())
 
 	// this amount is the overhead + tensors in memory
 	// TODO: get this from the llama.cpp's graph calculations instead of
 	// estimating it's 1/6 * kv_cache_size * num_gqa
 	graph := int64(ggml.KV().GQA()) * kv / 6
+	mem += graph
 
-	if slices.Contains(cpuOnlyFamilies, ggml.KV().Architecture()) {
+	if mem > vram {
 		opts.NumGPU = 0
 	}
 
-	info := gpu.GetGPUInfo()
-	switch runtime.GOOS {
-	case "darwin":
-		if opts.NumGPU == 0 {
-			break
-		}
+	if opts.NumGPU < 0 && info.Library != "cpu" {
```

Comment:
Maybe I'm misreading things, but shouldn't this be `opts.NumGPU <= 0`   otherwise combining line 70 and 94 doesn't that mean if the model won't completely fit we fallback to CPU?